### PR TITLE
feat(nocode): Add get entities v2 endpoint that can get without snapshot

### PR DESCRIPTION
**Pull request by dexter-mh-lee**
As part of the effort to make it easier to add new entities, we want to remove our dependency on snapshot and aspect unions. This is the first step to do so. 

Create a get entities v2 endpoint that fetches the aspects for the given set of urns and returns them in a dynamic fashion. The aspect is encoded into a placeholder PDL object called Aspect (which has no fields defined in the PDL). The caller of the get function can generate the actual aspect from the DataMap inside this placeholder. 

## Checklist
- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: entity-registry/src/main/java/com/linkedin/metadata/models/registry/ConfigEntityRegistry.java [hunk 1]
```diff
@@ -102,16 +102,17 @@ public ConfigEntityRegistry(DataSchemaFactory dataSchemaFactory, InputStream con
...
       aspectSpecs.add(getAspectSpec(entity.getKeyAspect(), entitySpecBuilder));
       entity.getAspects().forEach(aspect -> aspectSpecs.add(getAspectSpec(aspect, entitySpecBuilder)));
 
-      EntitySpec entitySpec = entitySpecBuilder.buildEntitySpec(entitySchema.get(), aspectSpecs);
+      EntitySpec entitySpec;
+      Optional<DataSchema> entitySchema = dataSchemaFactory.getEntitySchema(entity.getName());
+      if (!entitySchema.isPresent()) {
+        entitySpec = entitySpecBuilder.buildConfigEntitySpec(entity.getName(), entity.getKeyAspect(), aspectSpecs);
+      } else {
+        entitySpec = entitySpecBuilder.buildEntitySpec(entitySchema.get(), aspectSpecs);
```
Comment (jjoyce0510) [inline entity-registry/src/main/java/com/linkedin/metadata/models/registry/ConfigEntityRegistry.java:114]: Why is this necessary? Shouldn't config entity registry always build from config entity spec? Or is this to deal with the possibility that we've already found a snapshot?

File: metadata-io/src/main/java/com/linkedin/metadata/entity/EntityService.java [hunk 2]
```diff
@@ -133,6 +137,66 @@ protected EntityService(@Nonnull final EntityEventProducer producer, @Nonnull fi
...
+      @Nonnull final String entityName,
+      @Nonnull final Urn urn,
+      @Nonnull final String aspectName) throws Exception;
+
+  /**
+   * Retrieves the specific version of the aspect for the given urn
+   *
+   * @param entityName name of the entity to fetch
+   * @param urn urn to fetch
+   * @param aspectName name of the aspect to fetch
```
Comment (jjoyce0510) [inline metadata-io/src/main/java/com/linkedin/metadata/entity/EntityService.java:190]: Thank you for adding these comments!

File: metadata-io/src/main/java/com/linkedin/metadata/entity/ebean/EbeanEntityService.java [hunk 3]
```diff
@@ -142,6 +148,73 @@ public RecordTemplate getAspect(@Nonnull final Urn urn, @Nonnull final String as
...
+            Collectors.mapping(Map.Entry::getValue, Collectors.toList())));
+
+    // For each input urn, get the aspects corresponding to the urn. If empty, return the key aspect
+    final Map<Urn, List<EnvelopedAspect>> result = new HashMap<>();
+    for (Urn urn : urns) {
+      List<EnvelopedAspect> aspects = urnToAspects.getOrDefault(urn.toString(), Collections.emptyList());
+      EnvelopedAspect keyAspect = getKeyEnvelopedAspect(urn);
+      // Add key aspect if it does not exist in the returned aspects
+      if (aspects.isEmpty() || !aspectNames.contains(keyAspect.getName())) {
+        result.put(urn, ImmutableList.<EnvelopedAspect>builder().addAll(aspects).add(keyAspect).build());
```
Comment (jjoyce0510) [inline metadata-io/src/main/java/com/linkedin/metadata/entity/ebean/EbeanEntityService.java:192]: Actually unsure that we should do this. I think we'd previously had it but it seems kinda strange to inject an aspect that was not explicitly asked for. WDYT?

File: entity-registry/src/main/java/com/linkedin/metadata/models/registry/ConfigEntityRegistry.java (hunk 1 above)
Comment (dexter-mh-lee) [reply entity-registry/src/main/java/com/linkedin/metadata/models/registry/ConfigEntityRegistry.java:114]: I think the thought was to create the full entity spec if we have the schema for it. But I don't think it is necessary. Mostly backward compatibility

File: metadata-io/src/main/java/com/linkedin/metadata/entity/ebean/EbeanEntityService.java [hunk 4]
```diff
@@ -540,4 +612,50 @@ public ListUrnsResult listUrns(@Nonnull final String entityName, final int start
...
+      envelopedAspect.setCreated(new AuditStamp()
+          .setActor(Urn.createFromString(currAspectEntry.getCreatedBy()))
+          .setTime(currAspectEntry.getCreatedOn().getTime())
+      );
+      result.put(currKey, envelopedAspect);
+    }
+    return result;
+  }
+
+  private EnvelopedAspect getKeyEnvelopedAspect(final Urn urn) throws Exception {
```
Comment (jjoyce0510) [inline metadata-io/src/main/java/com/linkedin/metadata/entity/ebean/EbeanEntityService.java:645]: Almost feel like we should stop returning the Key Aspect in the case in which it's missing.. Injecting it just creates more confusion

File: metadata-io/src/test/java/com/linkedin/metadata/entity/EbeanEntityServiceTest.java [hunk 5]
```diff
@@ -277,6 +344,49 @@ public void testIngestGetLatestAspect() throws Exception {
...
+    assertTrue(DataTemplateUtil.areEqual(EbeanUtils.parseSystemMetadata(readEbean1.getSystemMetadata()), metadata1));
+
+    verify(_mockProducer, times(1)).produceMetadataAuditEvent(Mockito.eq(entityUrn), Mockito.eq(null), Mockito.any(),
+        Mockito.any(), Mockito.any(), Mockito.eq(MetadataAuditOperation.UPDATE));
+
+    verify(_mockProducer, times(1)).produceMetadataAuditEvent(Mockito.eq(entityUrn), Mockito.notNull(), Mockito.any(),
+        Mockito.any(), Mockito.any(), Mockito.eq(MetadataAuditOperation.UPDATE));
+
+    verifyNoMoreInteractions(_mockProducer);
+  }
```
Comment (jjoyce0510) [inline metadata-io/src/test/java/com/linkedin/metadata/entity/EbeanEntityServiceTest.java:388]: Awesome tests!

File: metadata-service/restli-servlet-impl/src/main/java/com/linkedin/metadata/resources/entity/EntityV2Resource.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+            String.format("Failed to get entity with urn: %s, aspects: %s", urn, projectedAspects), e);
+      }
+      return entities.get(urn);
+    }, MetricRegistry.name(this.getClass(), "get"));
+  }
+
+  @RestMethod.BatchGet
+  @Nonnull
+  @WithSpan
+  public Task<Map<Urn, EntityResponse>> batchGet(@Nonnull Set<String> urnStrs,
```
Comment (jjoyce0510) [inline metadata-service/restli-servlet-impl/src/main/java/com/linkedin/metadata/resources/entity/EntityV2Resource.java:71]: Just a note for us to discuss later: The downside to using restli batch get here is that there are specific requirements for the batch get method - like providing a list in the query parameters in a super Restli-specific way: GET /entitiesV2?ids=List(id1,id2) or something similar. If we started to use Spring directly we could actually design the API however we want. Definitely something worth considering

File: metadata-io/src/main/java/com/linkedin/metadata/entity/ebean/EbeanEntityService.java (hunk 4 above)
Comment (dexter-mh-lee) [reply metadata-io/src/main/java/com/linkedin/metadata/entity/ebean/EbeanEntityService.java:645]: Yeah this was something that bothered me. I returned it for backward compatibility, but seems very counter-intuitive

File: metadata-io/src/main/java/com/linkedin/metadata/entity/ebean/EbeanEntityService.java (hunk 3 above)
Comment (dexter-mh-lee) [reply metadata-io/src/main/java/com/linkedin/metadata/entity/ebean/EbeanEntityService.java:192]: Agreed. I wasn't sure if this would break things.

**Review by shirshanka — APPROVED** [review]
LGTM
